### ISSUE TITLE
Road to 1.0

### ISSUE THREAD

**Issue by yupferris** [yupferris/vb-rs#22]
I've been thinking about what Rustual Boy 1.0 looks like, and I think the following things are what would make a great first proper release:

- libretro/retroarch as the primary, full-featured frontend
- cli frontend for something more minimal and with a basic debugger
- 100% game compatibility (at least in playable status)
- Comparable or lower resource usage with other Virtual Boy emulators
- Some basic documentation

So, I'd like to track this as an issue to get it out of my head and also get some feedback on other things we could do for this. I think keeping this list as minimal as possible is necessary to keep it reachable, and I also think we're already pretty close!

In order to reach these points, I think some elaboration/tracking would be useful, so here's an attempt to enumerate/explain things in more detail. Note that I'm quite open to discussion about what these steps should look like, so feel free to discuss here and/or on discord, but at least I'd like to just lay something out initially to put a pin on what to discuss:

- [ ] libretro/retroarch as the primary, full-featured frontend
  - I don't think this means we need to support all of the platforms that libretro supports right off the bat, but Windows, Mac, and Linux should be a must.
  - [x] Initial frontend implementation (see https://github.com/sebbert/rustual-boy-libretro for current status/progress)
  - [x] Migrate frontend implementation into main repo
  - [ ] Refactor/fix up/mature frontend impl
    - See https://github.com/emu-rs/rustual-boy/issues/36 for more info
- [x] cli frontend for something more minimal and with a basic debugger
  - We're _almost_ there, except for some stability stuff on x86_64 on mac (see https://github.com/emu-rs/rustual-boy/issues/13 for more info).
- [x] 100% game compatibility
  - At least in playable status (see https://github.com/emu-rs/rustual-boy#known-game-compatibility for current status)
  - [x] Space Invaders Virtual Collection
    - Lazers (from both the player and the invaders) aren't visible in the main play area in Virtual 3D mode. Haven't investigated this specifically, but I suspect some CPU tests against the hardware wrt floats and bit strings might explain it.
  - [x] Nester's Funky Bowling
    - Screen goes black after initial warning screen and appears to softlock. Need to investigate further.
    - Turns out this was fixed in 4bc178e437cd734f1953ed7e648194aec579eca1, as well as the softlock bugs in Golf. I'm guessing VIP interrupts were being swallowed if the CPU couldn't handle them (i.e. if the CPU was already handling another VIP interrupt). This newer (simpler) behavior will fire a new VIP interrupt when the current one is finished in that case.
  - [x] Mario's Tennis
    - Severe graphical glitches when transitioning between screens. After a _lot_ of investigation, I'm thinking this has to do with a CPU emulation bug, but I don't know where. I'd like to make some program fuzzing tests to test for precise differences with the hardware.
    - Fixed in b648fb7c6e4eaa12adc1ccf58d6946116dc671a4. The issue ended up being that DPRST/XPRST bits of DPCTRL/XPCRTL not only clear interrupt pending bits, but interrupt enable bits as well. This means certain interrupts won't fire until explicitly enabled after these reset bits are set.
  - [x] Space Squash
    - Game doesn't start; seems to be waiting on an interrupt. Some previous investigation:
      - https://github.com/emu-rs/rustual-boy/commit/74235249a1abfca8d4b3d80e8c3c6b37230679a2
      - https://github.com/emu-rs/rustual-boy/commit/0ff9c61efb188832680292a11c1a24c5c4f25360
    - One thing we don't emulate that the hw does is a startup period where the display mirrors stabilize before the "display ready" bit is set. After trying a [very simple implementation](https://github.com/emu-rs/rustual-boy/commit/5cfed9e420bf8d645adabe4fc557dc39c0dbf527) it didn't seem to affect much, but I'd like to do more debugging here to figure this out.
    - Interestingly, if for some reason the hardware would enable the display once it became stable, the game appears to boot. However, I don't recall reading anything like this in any of the hardware docs. Should make a hardware test to see if this might be the case.
    - This ended up being an incorrect fix in https://github.com/emu-rs/rustual-boy/commit/0ff9c61efb188832680292a11c1a24c5c4f25360. The correct fix is in https://github.com/emu-rs/rustual-boy/commit/c39ef8afd9d875452aad586f4c91bacbc01ea507.
  - [x] Vertical Force
    - Powerups/enemies often disappear/reappear along their paths; game softlocks on first boss. Some investigation:
      - https://github.com/emu-rs/rustual-boy/commit/9e1ad57a658b8b4311bd695563ba95cda289d22d
      - https://gist.github.com/yupferris/f7a11c9bd48a69b1b8ff08719c1e5f9a
      - https://github.com/emu-rs/rustual-boy/commit/6bdcae816630f8ee79593a4b3fda7904bca620c9
    - In the end, all it took to fix this was a couple of sign extensions: https://github.com/emu-rs/rustual-boy/commit/09cb07879b9c47aea2332a50050dd91efc1b2700
  - [x] Implement bit string op's
    - Assuming this implementation doesn't introduce/uncover additional bugs, this should cover the following titles:
      - 3-D Tetris
      - Golf
      - Nester's Funky Bowling
      - Red Alarm
- [ ] Comparable or lower resource usage with other Virtual Boy emulators
  - For example, mednafen uses max ~4% cpu on my box; Rustual Boy is currently max ~9%. This is understandable, as we haven't really got at this much specifically, but I'd like to tackle this head-on (preferably after the libretro and compatibility issues are resolved in order to not break moving parts).
  - There are some places in the emu where we use `unsafe` code and pointers to avoid bounds checks. In certain places this is quite sensible (reading ROM after masking the address, for example), but I'd like to actually revert all of these changes and only add them back if performance profiling points to them being a bottleneck. Basically, I want to be a bit more rigorous here than I was before other people were contributing to the emu. :)
  - Before tackling this we should also try to agree on some good metrics, get some real data, and make concrete goals from those
- [ ] Some basic documentation
  - Content-wise, we're lacking quite a bit, esp. re the debugger, keymap, anything libretro-related, etc.
  - Format-wise, the markdown here on github is ok, but it would be nice to have something redistributable that would go out with the releases (currently we have [this readme](https://github.com/emu-rs/rustual-boy/blob/master/doc/README), but that's really just a placeholder).

With all of these, I think we're looking pretty good for a first proper version :)

**Comment by Dr-Emann** [comment]
Perhaps this should be a milestone? I'm not sure how all that works in github.

**Comment by yupferris** [comment]
I considered a milestone, but that would mean separating the other bits into separate tickets. While that might also make sense, I'd prefer to have it in one tracking ticket as that makes it easier (at least for me) to see the whole thing at once and have overview.
